DESIGN_GUIDELINES.md

```diff
@@ -0,0 +1,671 @@
+
+
+## Technical / design debt
+
+The Docker CLI evolved over time, which also means that the design inherited
+some design-choices from the past that may not have been the best choices (in
+hindsight), but cannot be changed without introducing breaking changes.
+
+This section describes some of these behaviors.
+
+### Legacy top level commands
+
+Historically, the Docker CLI had a limited set of commands, to manage images
+and containers (`docker run`, `docker pull`, `docker push`). With the introduction
+of other type of objects (volumes, networks, services, plugins), this pattern
+did not scale.
+
+No new top-level commands should be added; top-level commands are reserved for
+management commands going forward.
+
+> **Note**: some less-frequently used legacy top-level commands can be hidden by
+> setting the `DOCKER_HIDE_LEGACY_COMMANDS` environment variable. Setting this
+> variable _hides_ the commands, but they will still remain active for backward
+> compatibility.
+
+### Exit code for filtered results
+
+When filtering results, and no results were found, the CLI produces a zero (success)
+exit code. Reason for this is that the action was _successful_, but happened to
+not produce any matching results.
+
+For example;
+
+```bash
+docker container ls --filter status=exited
+
+CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
+
+echo $?
+0
+```
+
+```bash
+ls *.bla 2> /dev/null || echo "no such thing"
+no such thing
+```
+
+
+Producing a zero exit code can complicate using these commands in scripting
+situations, but has been considered too much of a breaking change to change (See
+[#27657](https://github.com/moby/moby/issues/27657#issuecomment-258271259) and
+[#28951](https://github.com/moby/moby/issues/28951)).
+
+### Single-value flags can be specified multiple times
+
+The Docker CLI accepts flags to be set multiple times, even if an option that's
+set through that flag accepts a single value. If a single-value flag is set
+multiple times, no error is produced, and latter values override prior values.
+
+For example, the following command runs successfully;
+
+```bash
+container create --name one --name two --name three busybox
+c872b39344646e86ef7d83f846e47f0cb92abc200938187f87052737026433d1
+
+echo $?
+0
+```
+
+And creates a container named `three`:
+
+```bash
+docker container inspect --format '{{.Name}}' c872b39344646e86ef7d83f846e47f0cb92abc200938187f87052737026433d1
+/three
+```
+
+> **Note**: it may be worth revisiting this situation, and only keep this
+> behavior for existing flags (for backward compatibility), and enforce single-
+> value flags to be specified only once going forward.
+
+### Passing input from `stdin`
+
+There is some inconsistency in notations used to pass input from `stdin`.
+
+```
+Usage:	docker config create [OPTIONS] CONFIG file|-
+
+```
+
+Which allows passing
+
+```bash
+echo $SOME_ENV_VAR | docker config create myconfig -
```

Comment:
I don't see the inconsistency here 🤔

Reply:
Need to read back what I had here; this might've just been some scratch notes I made that I still had to describe 😊